components/service/sync-autofill/src/main/java/mozilla/components/service/sync/autofill/DefaultCreditCardValidationDelegate.kt

```diff
@@ -0,0 +1,69 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.service.sync.autofill
+
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+import mozilla.components.concept.storage.CreditCard
+import mozilla.components.concept.storage.CreditCardEntry
+import mozilla.components.concept.storage.CreditCardNumber
+import mozilla.components.concept.storage.CreditCardValidationDelegate
+import mozilla.components.concept.storage.CreditCardValidationDelegate.Result
+import mozilla.components.concept.storage.CreditCardsAddressesStorage
+
+/**
+ * A delegate that will check against the [CreditCardsAddressesStorage] to determine if a given
+ * [CreditCard] can be persisted and returns information about why it can or cannot.
+ *
+ * @param storage An instance of [CreditCardsAddressesStorage].
+ */
+class DefaultCreditCardValidationDelegate(
+    private val storage: Lazy<CreditCardsAddressesStorage>
+) : CreditCardValidationDelegate {
+
+    private val coroutineContext by lazy { Dispatchers.IO }
+
+    override suspend fun validate(creditCard: CreditCardEntry): Result =
+        withContext(coroutineContext) {
+            val creditCards = storage.value.getAllCreditCards()
+
+            val foundCreditCard = if (creditCards.isEmpty()) {
+                // No credit cards exist in the storage, create a new credit card to the storage.
+                null
+            } else {
+                // Found a matching guid and credit card number -> update
+                creditCards.find {
```

Comment:
@mhammond brought up a case that this doesn't handle well - when a credit card expires, in Australia you'll get issued a brand new card with a new number. Current code will create a "dupe", since it does look like a brand new card.

Not much we can do about that at this level, this is more of a UX/product call - e.g. at some point, we could prompt a user if a new card should replace an old one, for example.